lib/tapioca/compilers/dsl_compiler.rb

```diff
@@ -0,0 +1,125 @@
+        @error_handler = error_handler
+      end
+
+      sig { params(blk: T.proc.params(constant: Module, rbi: String).void).void }
+      def run(&blk)
+        constants_to_process = gather_constants(requested_constants)
+
+        if constants_to_process.empty?
+          report_error(<<~ERROR)
+            No classes/modules can be matched for RBI generation.
+            Please check that the requested classes/modules include processable DSL methods.
+          ERROR
+        end
+
+        constants_to_process.sort_by { |c| c.name.to_s }.each do |constant|
+          rbi = rbi_for_constant(constant)
+          next if rbi.nil?
+
+          blk.call(constant, rbi)
+        end
+      end
+
+      private
+
+      sig { params(requested_generators: T::Array[String]).returns(Proc) }
+      def generator_filter(requested_generators)
+        return proc { true } if requested_generators.empty?
+
+        generators = requested_generators.map(&:downcase)
+
+        proc do |klass|
+          generator = klass.name&.sub(/^Tapioca::Compilers::Dsl::/, '')&.downcase
+          generators.include?(generator)
+        end
+      end
+
+      sig { params(requested_generators: T::Array[String]).returns(T::Enumerable[Dsl::Base]) }
+      def gather_generators(requested_generators)
+        generator_filter = generator_filter(requested_generators)
+
+        Dsl::Base.descendants.select(&generator_filter).map(&:new)
+      end
+
+      sig { params(requested_constants: T::Array[Module]).returns(T::Set[Module]) }
+      def gather_constants(requested_constants)
+        constants = generators.map(&:processable_constants).reduce(Set.new, :union)
+        constants &= requested_constants unless requested_constants.empty?
+        constants
+      end
+
+      sig { params(constant: Module).returns(T.nilable(String)) }
+      def rbi_for_constant(constant)
+        parlour = Parlour::RbiGenerator.new(sort_namespaces: true)
+
+        generators.each do |generator|
+          next unless generator.handles?(constant)
+          generator.decorate(parlour.root, constant)
+        end
+
+        return if parlour.root.children.empty?
+
+        resolve_conflicts(parlour)
+
+        parlour.rbi("true").strip
+      end
+
+      sig { params(parlour: Parlour::RbiGenerator).void }
+      def resolve_conflicts(parlour)
+        Parlour::ConflictResolver.new.resolve_conflicts(parlour.root) do |msg, candidates|
+          error = StringIO.new
+          error.puts "=== Error ==="
+          error.puts msg
+          error.puts "# Candidates"
+          candidates.each_with_index do |candidate, index|
+            error.puts "  #{index}. #{candidate.describe}"
+          end
+          report_error(error.to_s)
+        end
+      end
+
+      sig { params(error: String).returns(T.noreturn) }
+      def report_error(error)
+        handler = error_handler
+        if handler
+          handler.call(error)
+        else
+          $stderr.puts(error)
```

Comment:
Nitpick: It doesn't feel nice to call `stderr` directly when we have `say_error` but I guess we don't want to include it in this class for one call?

Reply:
Yes, this is supposed to be the ultimate fallback when no error handler was registered but there was still an error in processing. That is not supposed to happen since the only consumer of this class is the `Generator` and the handler is supplied there anyway, but I just wanted to be safer in the implementation.